### ISSUE TITLE
RC4 encryption algorithm is not supported

### ISSUE THREAD

**Issue by kurtseebauer** [sijms/go-ora#345]
Hi all, 

I'm getting an error trying to connect to an oracle 19 with TCPS/SSL. 

go-ora version: v2.6.16

code: 
```
// options
	if protocol == "TCPS" {
		urlOptions["TRACE FILE"] = "trace.log"
		urlOptions["SSL"] = "true"
		urlOptions["SSL Verify"] = "false"
	}

// build the db connection
	connStr := fmt.Sprintf(`(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=%v)(HOST=%v)(PORT=%v)))(CONNECT_DATA=(SID=%v)))`, protocol, host, port, sid)

       databaseUrl = goora.BuildJDBC(user, password, connStr, urlOptions)

	db, err := sqlx.Open("oracle", databaseUrl)
	if err != nil {
		panic(err.Error())
	}

	sqldb, err := sql.Open("oracle", databaseUrl)
	if err != nil {
		panic(err.Error())
	}
        // works this far, but both Ping() versions fail:
	sqldb.Ping()

	err = db.Ping()
	if err != nil {
		panic(err.Error())
	}
```
where `PROTOCOL` is `TCPS`.

The trace file shows that we just send one package with the connection string to the db and we get an error in the Ping() call. The Open() calls work fine. 

```
2023-04-13T11:32:26.9428: Connect
2023-04-13T11:32:26.9438: using: 10.redacted:1531 ..... [SUCCEED]
2023-04-13T11:32:26.9443: Using SSL/TLS
2023-04-13T11:32:26.9449: Open :(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCPS)(HOST=redacted)(PORT=1531)))(CONNECT_DATA=(SID=TESTSID)))
2023-04-13T11:32:26.9460: 
Write packet:
00000000  00 b7 00 00 01 00 00 00  01 3d 01 2c 0c 01 ff ff  |.........=.,....|
00000010  ff ff 4f 98 00 00 00 01  00 71 00 46 00 00 00 00  |..O......q.F....|
00000020  01 01 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000030  00 00 00 00 00 00 00 00  00 00 00 20 00 00 00 20  |........... ... |
00000040  00 00 00 00 00 00 28 44  45 53 43 52 49 50 54 49  |......(DESCRIPTI|
00000050  4f 4e 3d 28 41 44 44 52  45 53 53 5f 4c 49 53 54  |ON=(ADDRESS_LIST|
00000060  3d 28 41 44 44 52 45 53  53 3d 28 50 52 4f 54 4f  |=(ADDRESS=(PROTO|
00000070  43 4f 4c 3d 54 43 50 53  29 28 48 4f 53 54 3d 31  |COL=TCPS)(HOST=1|
00000080  30 2e 31 37 33 2e 32 30  38 2e 37 31 29 28 50 4f  |0....)(PO|
00000090  52 54 3d 31 35 33 31 29  29 29 28 43 4f 4e 4e 45  |RT=1531)))(CONNE|
000000a0  43 54 5f 44 41 54 41 3d  28 53 49 44 3d 54 45 53  |CT_DATA=(SID=TES|
000000b0  54 4c 4f 4b 29 29 29                              |TSID)))|

```
(that's all in the file).

The error returned from the Ping() call is `EOF`. 

Any help would be very appreciated. Thank you in advance!

**Comment by sijms** [comment]
pass wallet location and test again

**Comment by kurtseebauer** [comment]
Hi @sijms , thx for the quick reply. I only have username/password for this database, so I can't test that directly.
Do you mean that a wallet in the client is absolutely necessary for connecting? 

https://blogs.oracle.com/developers/post/securely-connecting-to-autonomous-db-without-a-wallet-using-tls#Enable%20TLS
seems to suggest the it depends on the server config, with either mTLS (mutual TLS with certificates on both ends) required or TLS permitted, is that right?

Thank you for your time!

**Comment by sijms** [comment]
I still have problem creating account on oracle cloud
would you share the database URL without username and password
I find similar issue #299 
according to that. the issue related to calling `session.negotiate()` which is necessary in old way `mTLS` 
if you can put a breakpoint at line# 602 `network/session.go` and see if the driver execute line 603 or not
also at line# 667 same file
if `session.negotiate()` is execute just comment it and test again

**Comment by kurtseebauer** [comment]
Hi @sijms, thanks for your effort. 

To clear some things up, I'm not working with the autonomous oracle, it is just the piece of documentation I found and I don't know if the same applies to a regular (self-hosted) DB. 
The program I'm writing runs on a couple of customer sites which I have limited access to, until they started upgrading to oracle 19. I can only copy my binary to one particular site and run tests without debugging. I asked one of the DB admins for information about the server config (no response so far). 

What I understood from the above blog post and 
https://docs.oracle.com/en/database/oracle/oracle-database/19/dbseg/configuring-secure-sockets-layer-authentication.html#GUID-6AD89576-526F-4D6B-A539-ADF4B840819F
it all depends on whether in sqlnet.ora the settings SSL_CLIENT_AUTHENTICATION is set to true (mTLS -> wallet needed) or false (wallet not necessary). 

So the solution _may_ just be that we need a wallet for every future customer, which is an effort that adds up after a few dozen onboardings. Or it _may_ be possible that a wallet is not needed, depending on the config. 

So far I managed without installing a local oracle instance (and messing with the config), but it looks a lot like I have to now.  Lots of knowledge gaps I need to fill.

**Comment by sijms** [comment]
Hi @kurtseebauer 
if you are not using autonomous database so it is likely that server using mTLS not TLS
client authentication is supported in the driver and it is ability of the user to login into server without username and password only using client certificate and it need oracle wallet to work
for TCPS with mTLS you should have a `client certificate` connected to `client DNS`. then this certificate should appear in `server wallet trusted certificate` before connection can be accepted by the server. so you should have wallet to connect
if `client DNS` changed for any reason you should re-create client certificate again and re-define it in `server trusted certificate`

with the above items completed connection will pass smooth between server and client
also the above is tested against oracle server 19c with no issue

**Comment by cjbj** [comment]
> I still have problem creating account on oracle cloud

@sijms Mail me if you still have issues.

Chris

**Comment by kurtseebauer** [comment]
So, I got new info from the customer. It looks like I was sent down the wrong rabbit hole with TCPS altogether.

The database config according to the db admin is:
```
SQLNET.CRYPTO_CHECKSUM_TYPES_SERVER = (MD5)
SQLNET.CRYPTO_CHECKSUM_SERVER = REQUIRED
SQLNET.ENCRYPTION_TYPES_SERVER = (rc4_256)
SQLNET.ENCRYPTION_SERVER = REQUIRED
```
And we failed to connect using TCP. 

From `encrypt_service.go` it looks like RC4_256 is not supported as an encrpytion algorithm, is that right @sijms ?
```
	output := &encryptService{
		defaultService: defaultService{
			comm:        comm,
			serviceType: 2,
			version:     0xB200200,
			availableServiceNames: []string{"", "RC4_40", "RC4_56", "RC4_128", "RC4_256",
				"DES40C", "DES56C", "3DES112", "3DES168", "AES128", "AES192", "AES256"},
			availableServiceIDs: []int{0, 1, 8, 10, 6, 3, 2, 11, 12, 15, 16, 17},
		},
	}
	err := output.buildServiceList([]string{"DES56C", "AES128", "AES192", "AES256"}, true, true)
```
Then the question would be, is there a chance we can support RC4_256 or do I have to use a different driver? :-|

**Comment by sijms** [comment]
I will try to add it

**Comment by kurtseebauer** [comment]
I did a bit of an investigation how RC4 works in the jdbc driver: https://github.com/sijms/go-ora/pull/354
@sijms do you have some good/better sources to figure out how oracle uses RC4? I'm kinda stuck there.

**Comment by sijms** [comment]
I am working on it

**Comment by sijms** [comment]
fixed in v2.7.3
